src/runtime_src/core/pcie/driver/linux/xocl/subdev/accel_deadlock_detector.c

```diff
@@ -0,0 +1,288 @@
+    if (!res) {
+        err = -ENOMEM;
+        xocl_err(&pdev->dev, "platform_get_resource failed for accel_deadlock_detector_probe , err (-ENOMEM): %d", err);
+        goto done;
+    }
+
+
+    xocl_info(&pdev->dev, "IO start: 0x%llx, end: 0x%llx",
+                res->start, res->end);
+
+    accel_deadlock_detector->base = ioremap_nocache(res->start, res->end - res->start + 1);
+    if (!accel_deadlock_detector->base) {
+        err = -EIO;
+        xocl_err(&pdev->dev, "Map iomem failed");
+        goto done;
+    }
+
+    accel_deadlock_detector->start_paddr = res->start;
+    accel_deadlock_detector->range = res->end - res->start + 1;
+
+    err = sysfs_create_group(&pdev->dev.kobj, &accel_deadlock_detector_attr_group);
+    if (err) {
+        xocl_err(&pdev->dev, "create accel_deadlock_detector sysfs attrs failed: %d", err);
+    }
+
+done:
+    if (err) {
+        accel_deadlock_detector_remove(pdev);
+        return err;
+    }
+    return 0;
+}
+
+static int accel_deadlock_detector_open(struct inode *inode, struct file *file)
+{
+    struct xocl_accel_deadlock_detector *accel_deadlock_detector = NULL;
+
+    accel_deadlock_detector = xocl_drvinst_open_single(inode->i_cdev);
+    if (!accel_deadlock_detector)
+        return -ENXIO;
+    file->private_data = accel_deadlock_detector;
+    return 0;
+}
+
+static int accel_deadlock_detector_close(struct inode *inode, struct file *file)
+{
+    struct xocl_accel_deadlock_detector *accel_deadlock_detector = file->private_data;
+
+    xocl_drvinst_close(accel_deadlock_detector);
+    return 0;
+}
+
+long accel_deadlock_detector_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+    struct xocl_accel_deadlock_detector *accel_deadlock_detector;
+    void __user *data;
+    long result = 0;
+
+    accel_deadlock_detector = (struct xocl_accel_deadlock_detector *)filp->private_data;
+    data = (void __user *)(arg);
+
+    mutex_lock(&accel_deadlock_detector->lock);
+
+    switch (cmd) {
+        case ACCEL_DEADLOCK_DETECTOR_IOC_RESET:
+            break;
+        case ACCEL_DEADLOCK_DETECTOR_IOC_GET_STATUS:
+            result = XOCL_READ_REG32(accel_deadlock_detector->base + 0x0);
+            break;
+        default:
+            result = -ENOTTY;
+    }
+    mutex_unlock(&accel_deadlock_detector->lock);
+
+    return result;
+}
+
+
+static int accel_deadlock_detector_mmap(struct file *filp, struct vm_area_struct *vma)
+{
+
+    int rc;
+    unsigned long off;
+    unsigned long phys;
+    unsigned long vsize;
+    unsigned long psize;
+    struct xocl_accel_deadlock_detector *accel_deadlock_detector = (struct xocl_accel_deadlock_detector *)filp->private_data;
+    BUG_ON(!accel_deadlock_detector);
+
+    off = vma->vm_pgoff << PAGE_SHIFT;
+    /* BAR physical address */
+    phys = accel_deadlock_detector->start_paddr + off;
+    vsize = vma->vm_end - vma->vm_start;
+    /* complete resource */
+    psize = accel_deadlock_detector->range - off;
```

Comment:
what if off > range?

Reply:
I discussed this with Chien-Wei. He mentioned that we can add a sanity check for safety. However, I found that we will need this check for all the profile monitor subdevices too. I can open a new pull request with the same fix for all the subdevices together.

Reply:
Sounds great to me. Thanks.